File: pentomino/pentomino.py
import string
from copy import copy
from heapq import heappush, heappop
import random
import time
from typing import Iterator, List

PENTO = 5
_ = None

# O to Z
piece_labels = string.ascii_uppercase[-12:]

piece_shape = {
    "O": [["@", "@", "@", "@", "@"]],
    "P": [["@", "@", "@"],
          ["@", "@"]],
    "Q": [["@", "@", "@", "@"],
          ["@"]],
    "R": [["@", "@"],
          [  _, "@", "@"],
          [  _, "@"]],
    "S": [[  _, "@", "@", "@"],
          ["@", "@"]],
    "T": [["@"],
          ["@", "@", "@"],
          ["@"]],
    "U": [["@", "@", "@"],
          ["@",   _, "@"]],
    "V": [["@", "@", "@"],
          ["@"],
          ["@"]],
    "W": [[  _, "@", "@"],
          ["@", "@"],
          ["@"]],
    "X": [[  _, "@"],
          ["@", "@", "@"],
          [  _, "@"]],
    "Y": [["@", "@", "@", "@"],
          [  _, "@"]],
    "Z": [["@", "@"],
          [  _, "@"],
          [  _, "@", "@"]],
}


class MinoPosition:
    def __init__(self, minos):
        self.minos = minos

    def turn_flip(self, factor):
        """
        Transforms the piece's mino coordinates.
        factor 0-3: rotations
        factor 4-7: flipped then rotations
        """
        new_position = []
        for y, x in self.minos:
            if factor == 0: # 0 deg
                ny, nx = y, x
            elif factor == 1: # 90 deg
                ny, nx = x, -y
            elif factor == 2: # 180 deg
                ny, nx = -y, -x
            elif factor == 3: # 270 deg
                ny, nx = -x, y
            elif factor == 4: # flipped + 0 deg
                ny, nx = y, -x
            elif factor == 5: # flipped + 90 deg
                ny, nx = -x, -y
            elif factor == 6: # flipped + 180 deg
                ny, nx = -y, x
            elif factor == 7: # flipped + 270 deg
                ny, nx = x, y
            new_position.append((ny, nx))
        new_position.sort()
        self.minos = new_position

    def move(self, basecell, y, x):
        y0, x0 = self.minos[basecell]
        self.minos = [(cy - y0 + y, cx - x0 + x) for cy, cx in self.minos]

    def __repr__(self):
        return f"MinoPosition({repr(self.minos)})"

    def __iter__(self):
        return iter(self.minos)


class Board:
    def __init__(self, ny, nx):
        self.cells = [[None] * nx for _ in range(ny)]
        self.put_history = []

    def span(self):
        ny = len(self.cells)
        nx = len(self.cells[0])
        return (ny, nx)

    def is_empty_at(self, y, x):
        ny, nx = self.span()
        return 0 <= y < ny and 0 <= x < nx and self.cells[y][x] is None

    def can_accept(self, minos: MinoPosition) -> bool:
        return all(self.is_empty_at(y, x) for y, x in minos)

    def put(self, position: MinoPosition, k: str):
        for y, x in position:
            self.cells[y][x] = k
        self.put_history.append(k)

    def remove(self, position: MinoPosition):
        """Removes a piece from the board, undoing a `put` operation."""
        for y, x in position:
            self.cells[y][x] = None
        self.put_history.pop()

    def print(self):
        for row in self.cells:
            plain_string = " ".join(c if c else "." for c in row)
            print(color_string(plain_string))


def color_string(plain_string):
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    LIGHT_RED = '\033[91m'
    LIGHT_GREEN = '\033[92m'
    LIGHT_YELLOW = '\033[93m'
    LIGHT_BLUE = '\033[94m'
    LIGHT_MAGENTA = '\033[95m'
    LIGHT_CYAN = '\033[96m'
    RESET = '\033[0m'
    color_map = {
        'O': RED,
        'P': GREEN,
        'Q': YELLOW,
        'R': BLUE,
        'S': MAGENTA,
        'T': CYAN,
        'U': LIGHT_RED,
        'V': LIGHT_GREEN,
        'W': LIGHT_YELLOW,
        'X': LIGHT_BLUE,
        'Y': LIGHT_MAGENTA,
        'Z': LIGHT_CYAN
    }
    colored_string = []
    for c in plain_string:
        if c in color_map:
            code = color_map[c]
            colored_string.extend([code, c, RESET])
            continue
        colored_string .append(c)
    return ''.join(colored_string)


class Piece:
    def __init__(self, label):
        self.label = label
        shape = piece_shape[label]
        minos = []
        for y, row in enumerate(shape):
            for x, blk in enumerate(row):
                if blk:
                    minos.append((y, x))
        self.minos = minos

    def __eq__(self, other):
        return self.label == other.label

    def __str__(self):
        return self.label

    def __hash__(self):
        return hash(self.label)

    def possible_positions(self, y, x, board: Board) -> Iterator[MinoPosition]:
        # randomize piece direction
        turn_factors = [i for i in range(0, 8)]
        random.shuffle(turn_factors)
        cell_indexes = [i for i in range(0, PENTO)]
        random.shuffle(cell_indexes)

        done = set()
        for factor in turn_factors:
            for center_cell in cell_indexes:
                position = MinoPosition(self.minos)
                position.turn_flip(factor)
                position.move(center_cell, y, x)

                # for deduplicating symmetry
                key = tuple(position.minos)
                if key in done:
                    continue
                done.add(key)

                if board.can_accept(position):
                    yield position


class NoAnswer(Exception):
    pass


class FitStatistics:
    def __init__(self, pieces: List[Piece]):
        self.pieces = pieces
        # start from 1 to avoid zero-division error
        self.trial = {p: 1 for p in self.pieces}
        self.fit = {p: 1 for p in self.pieces}

    def countup_trial(self, piece: Piece):
        self.trial[piece] += 1

    def countup_fit(self, piece: Piece):
        self.fit[piece] += 1

    def fit_rate(self, piece: Piece):
        return self.fit[piece] / self.trial[piece]

    def list_fit_rate(self):
        rate_list = [(p, self.fit_rate(p)) for p in self.pieces]
        rate_list.sort(key=lambda t: t[1])
        return rate_list

    def mean_fit_rate(self):
        return sum(self.fit.values()) / sum(self.trial.values())


class SearchStatistics:
    def __init__(self, pieces: List[Piece]):
        self.reached_depth_count = {i: 0 for i in range(1, len(pieces) + 1)}

    def countup(self, reached_depth: int):
        if reached_depth > 0:
            self.reached_depth_count[reached_depth] += 1

    def total_trial(self):
        return sum(self.reached_depth_count.values())

    def mean_reached_depth(self):
        dc = self.reached_depth_count
        total_trials = self.total_trial()
        if not total_trials:
            return 0
        return sum(d * c for d, c in dc.items()) / total_trials

    def reach_distribution(self):
        total_trials = self.total_trial()
        if not total_trials:
            return []
        reachd = [
            (d, c / total_trials) for d, c in self.reached_depth_count.items()
        ]
        reachd.sort()
        return reachd


# Strategies of piece choice for the next fit trial
class FitDifficultPieceChoice:
    """A: Try from fit-diffucult piece"""
    def __init__(self, fit_statistics: FitStatistics):
        self.fit_statistics = fit_statistics

    def pop(self, pieces):
        pieces.sort(key=self.fit_statistics.fit_rate, reverse=True)
        return pieces.pop()


class FitEasyPieceChoice:
    """B: Try from fit-easy piece"""
    def __init__(self, fit_statistics: FitStatistics):
        self.fit_statistics = fit_statistics

    def pop(self, pieces):
        pieces.sort(key=self.fit_statistics.fit_rate, reverse=False)
        return pieces.pop()


class RandomPieceChoice:
    """C: Try randomly"""
    def __init__(self, fit_statistics: FitStatistics):
        pass

    def pop(self, pieces):
        random.shuffle(pieces)
        return pieces.pop()


def prioritize(open_edges, y, x, yspan, xspan):
    """Return priority value. Smaller value has higher priority."""
    ran = random.randrange(9999)
    # 1: short spanwise scan
    # return (y, x) if xspan < yspan else (x, y)

    # 2: open edges
    # return (open_edges, ran)

    # 3: open edges + short spanwise scan
    return (open_edges, y, x) if xspan < yspan else (open_edges, x, y)

    # 4: open edges + distance from two wall
    # return (open_edges, min(y, yspan - y) + min(x, xspan - x), ran)

    # 5: open edges + distance from nearest wall
    # return (open_edges, min(y, yspan - y, x, xspan - x), ran)


class Solver:
    """
    Encapsulates the pentomino puzzle solving logic.
    """
    def __init__(self, board: Board, pieces: List[Piece], strategy_class):
        self.board = board
        self.pieces = pieces
        self.fit_statistics = FitStatistics(self.pieces)
        self.search_statistics = SearchStatistics(self.pieces)
        self.piece_choice_strategy = strategy_class(self.fit_statistics)

    def solve(self):
        """
        Starts the backtracking solver.
        Returns the solved board or raises NoAnswer if no solution is found.
        """
        # (priority, y, x)
        cell_queue = [(0, 0, 0)]
        return self._backtrack(cell_queue, self.pieces)

    def _backtrack(self, cell_queue, remaining_pieces):
        while cell_queue:
            _, y, x = heappop(cell_queue)

            if not self.board.is_empty_at(y, x):
                continue

            pieces_not_tried = copy(remaining_pieces)

            while pieces_not_tried:
                piece = self.piece_choice_strategy.pop(pieces_not_tried)
                self.fit_statistics.countup_trial(piece)

                for piece_position in piece.possible_positions(y, x, self.board):
                    self.fit_statistics.countup_fit(piece)

                    self.board.put(piece_position, piece.label)
                    remaining_pieces2 = copy(remaining_pieces)
                    remaining_pieces2.remove(piece)

                    if not remaining_pieces2:
                        return self.board # Solved

                    cell_queue2 = copy(cell_queue)
                    four = ((-1, 0), (1, 0), (0, -1), (0, 1))

                    neighbors = {
                        (y1 + y2, x1 + x2)
                        for y1, x1 in four
                        for y2, x2 in piece_position
                        if self.board.is_empty_at(y1 + y2, x1 + x2)
                    }

                    for nby, nbx in neighbors:
                        parameters = {
                            "open_edges": sum(
                                self.board.is_empty_at(nby + y1, nbx + x1) for y1, x1 in four
                            ),
                            "y": y,
                            "x": x,
                            "yspan": self.board.span()[0],
                            "xspan": self.board.span()[1],
                        }
                        priority = prioritize(**parameters)
                        heappush(cell_queue2, (priority, nby, nbx))

                    try:
                        return self._backtrack(cell_queue2, remaining_pieces2)
                    except NoAnswer:
                        self.board.remove(piece_position)
                        continue

            self.search_statistics.countup(len(self.board.put_history))
            self.board.cells[y][x] = "+" # Mark as tried and failed

            if self.search_statistics.total_trial() % 1000 == 0:
                print(
                    "#search trial:",
                    self.search_statistics.total_trial(),
                    ":",
                    "".join(self.board.put_history),
                )
                self.board.print()
                print()
            
            self.board.cells[y][x] = None # Unmark

            raise NoAnswer()

        return self.board


YSPAN = 6
XSPAN = 10

if __name__ == "__main__":
    board = Board(YSPAN, XSPAN)
    pieces = [Piece(s) for s in piece_labels]
    random.shuffle(pieces)

    # for y, x in [(3,3), (3,4), (4,3), (4,4)]:
    #    board.cells[y][x] = "#"
    
    print("Initial board:")
    board.print()
    print()

    # Choose a strategy
    strategy = FitDifficultPieceChoice
    # strategy = FitEasyPieceChoice
    # strategy = RandomPieceChoice

    solver = Solver(board, pieces, strategy)

    start_time_ns = time.time_ns()

    try:
        answer_board = solver.solve()
        end_time_ns = time.time_ns()

        print("#fit rate")
        for piece, rate in solver.fit_statistics.list_fit_rate():
            print(f"{piece}: {rate:.3f}")

        print("#reach distribution")
        for depth, rate in solver.search_statistics.reach_distribution():
            print(f"{depth:2d}: {rate:.4f}")

        print()
        elapsed = (end_time_ns - start_time_ns) / 10**9
        print(f"elapsed time [s]: {elapsed:.2f}")
        print("total search trial:", solver.search_statistics.total_trial())
        fit_rate = solver.fit_statistics.mean_fit_rate()
        print(f"mean fit rate: {fit_rate:.2f}")
        depth = solver.search_statistics.mean_reached_depth()
        print(f"mean reached depth: {depth:.2f}")
        print()
        print("Solution:")
        answer_board.print()

    except NoAnswer:
        end_time_ns = time.time_ns()
        print("No solution found.")
        elapsed = (end_time_ns - start_time_ns) / 10**9
        print(f"elapsed time [s]: {elapsed:.2f}")

File: pentomino/Pentomino.java


import java.awt.Point;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Random;
import java.util.Set;

public class Pentomino {

    private static final int YSPAN = 6;
    private static final int XSPAN = 10;
    private static final int PENTO = 5;

    public static void main(String[] args) {
        Board board = new Board(YSPAN, XSPAN);
        board.print();

        List<Piece> pieces = new ArrayList<>();
        for (PieceLabel label : PieceLabel.values()) {
            pieces.add(new Piece(label));
        }
        Collections.shuffle(pieces);

        long startTime = System.nanoTime();

        PriorityQueue<int[]> cellQueue = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        cellQueue.add(new int[]{0, 0, 0}); // priority, y, x

        FitStatistics fitStatistics = new FitStatistics(pieces);
        SearchStatistics searchStatistics = new SearchStatistics();
        PieceChoiceStrategy pieceChoiceStrategy = new FitDifficultPieceChoice();

        try {
            Board answer = backtrack(cellQueue, pieces, board, fitStatistics, searchStatistics, pieceChoiceStrategy);
            long endTime = System.nanoTime();

            System.out.println("#fit rate");
            for (Map.Entry<PieceLabel, Double> entry : fitStatistics.listFitRate()) {
                System.out.printf("%s: %.3f%n", entry.getKey(), entry.getValue());
            }

            System.out.println("\n#reach distribution");
            for (Map.Entry<Integer, Double> entry : searchStatistics.reachDistribution()) {
                System.out.printf("%2d: %.4f%n", entry.getKey(), entry.getValue());
            }

            System.out.println();
            double elapsed = (endTime - startTime) / 1e9;
            System.out.printf("elapsed time [s]: %.2f%n", elapsed);
            System.out.println("total search trial: " + searchStatistics.totalTrial());
            System.out.printf("mean fit rate: %.2f%n", fitStatistics.meanFitRate());
            System.out.printf("mean reached depth: %.2f%n", searchStatistics.meanReachedDepth());
            System.out.println();
            answer.print();

        } catch (NoAnswer e) {
            System.out.println("No solution found.");
        }
    }

    private static Board backtrack(
            PriorityQueue<int[]> cellQueue,
            List<Piece> remainingPieces,
            Board board,
            FitStatistics fitStatistics,
            SearchStatistics searchStatistics,
            PieceChoiceStrategy pieceChoiceStrategy
    ) throws NoAnswer {
        while (!cellQueue.isEmpty()) {
            int[] cell = cellQueue.poll();
            int y = cell[1];
            int x = cell[2];

            if (!board.isEmptyAt(y, x)) {
                continue;
            }

            List<Piece> piecesNotTried = new ArrayList<>(remainingPieces);

            while (!piecesNotTried.isEmpty()) {
                Piece piece = pieceChoiceStrategy.pop(piecesNotTried, fitStatistics);
                fitStatistics.countupTrial(piece.getLabel());

                for (MinoPosition piecePosition : piece.possiblePositions(y, x, board)) {
                    fitStatistics.countupFit(piece.getLabel());

                    Board board2 = board.deepCopy();
                    board2.put(piecePosition, piece.getLabel());

                    List<Piece> remainingPieces2 = new ArrayList<>(remainingPieces);
                    remainingPieces2.remove(piece);

                    PriorityQueue<int[]> cellQueue2 = new PriorityQueue<>(cellQueue);
                    int[][] four = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

                    Set<Point> neighbors = new HashSet<>();
                    for (Point p : piecePosition.getMinos()) {
                        for (int[] dir : four) {
                            Point neighbor = new Point(p.x + dir[1], p.y + dir[0]);
                            if (board2.isEmptyAt(neighbor.y, neighbor.x)) {
                                neighbors.add(neighbor);
                            }
                        }
                    }

                    for (Point neighbor : neighbors) {
                        int openEdges = 0;
                        for (int[] dir : four) {
                            if (board2.isEmptyAt(neighbor.y + dir[0], neighbor.x + dir[1])) {
                                openEdges++;
                            }
                        }
                        cellQueue2.add(new int[]{prioritize(openEdges, neighbor.y, neighbor.x, board2.getNy(), board2.getNx()), neighbor.y, neighbor.x});
                    }

                    try {
                        return backtrack(cellQueue2, remainingPieces2, board2, fitStatistics, searchStatistics, pieceChoiceStrategy);
                    } catch (NoAnswer e) {
                        // continue
                    }
                }
            }

            searchStatistics.countup(board.putHistory.size());
            if (searchStatistics.totalTrial() % 1000 == 0) {
                System.out.println("#search trial: " + searchStatistics.totalTrial() + ": " + board.putHistory);
                board.print();
                System.out.println();
            }

            throw new NoAnswer();
        }

        return board;
    }

    private static int prioritize(int openEdges, int y, int x, int yspan, int xspan) {
        return openEdges;
    }

    static class NoAnswer extends Exception {
    }
}

enum PieceLabel {
    O, P, Q, R, S, T, U, V, W, X, Y, Z
}

class PieceShape {
    public static final Map<PieceLabel, int[][]> shapes = new EnumMap<>(PieceLabel.class);

    static {
        shapes.put(PieceLabel.O, new int[][]{{'@', '@', '@', '@', '@'}});
        shapes.put(PieceLabel.P, new int[][]{{'@', '@', '@'}, {'@', '@', 0}});
        shapes.put(PieceLabel.Q, new int[][]{{'@', '@', '@', '@'}, {'@', 0, 0, 0}});
        shapes.put(PieceLabel.R, new int[][]{{'@', '@', 0}, {0, '@', '@'}, {0, '@', 0}});
        shapes.put(PieceLabel.S, new int[][]{{0, '@', '@', '@'}, {'@', '@', 0, 0}});
        shapes.put(PieceLabel.T, new int[][]{{'@', 0, 0}, {'@', '@', '@'}, {'@', 0, 0}});
        shapes.put(PieceLabel.U, new int[][]{{'@', '@', '@'}, {'@', 0, '@'}});
        shapes.put(PieceLabel.V, new int[][]{{'@', '@', '@'}, {'@', 0, 0}, {'@', 0, 0}});
        shapes.put(PieceLabel.W, new int[][]{{0, '@', '@'}, {'@', '@', 0}, {'@', 0, 0}});
        shapes.put(PieceLabel.X, new int[][]{{0, '@', 0}, {'@', '@', '@'}, {0, '@', 0}});
        shapes.put(PieceLabel.Y, new int[][]{{'@', '@', '@', '@'}, {0, '@', 0, 0}});
        shapes.put(PieceLabel.Z, new int[][]{{'@', '@', 0}, {0, '@', 0}, {0, '@', '@'}});
    }
}

class MinoPosition {
    private List<Point> minos;

    public MinoPosition(List<Point> minos) {
        this.minos = new ArrayList<>(minos);
    }

    public void turnFlip(int factor) {
        List<Point> newPosition = new ArrayList<>();
        for (Point p : minos) {
            int y = p.y;
            int x = p.x;
            int ny, nx;
            switch (factor) {
                case 0: ny = y; nx = x; break;
                case 1: ny = x; nx = -y; break;
                case 2: ny = -y; nx = -x; break;
                case 3: ny = -x; nx = y; break;
                case 4: ny = y; nx = -x; break;
                case 5: ny = -x; nx = -y; break;
                case 6: ny = -y; nx = x; break;
                case 7: ny = x; nx = y; break;
                default: throw new IllegalArgumentException("Invalid factor");
            }
            newPosition.add(new Point(nx, ny));
        }
        newPosition.sort(Comparator.comparingInt((Point p) -> p.y).thenComparingInt(p -> p.x));
        this.minos = newPosition;
    }

    public void move(int baseCell, int y, int x) {
        Point p0 = minos.get(baseCell);
        List<Point> newMinos = new ArrayList<>();
        for (Point p : minos) {
            newMinos.add(new Point(p.x - p0.x + x, p.y - p0.y + y));
        }
        this.minos = newMinos;
    }

    public List<Point> getMinos() {
        return minos;
    }

    @Override
    public String toString() {
        return "MinoPosition(" + minos + ")";
    }
}

class Board {
    PieceLabel[][] cells;
    final List<PieceLabel> putHistory = new ArrayList<>();

    public Board(int ny, int nx) {
        cells = new PieceLabel[ny][nx];
    }

    public int getNy() {
        return cells.length;
    }

    public int getNx() {
        return cells[0].length;
    }

    public boolean isEmptyAt(int y, int x) {
        return y >= 0 && y < getNy() && x >= 0 && x < getNx() && cells[y][x] == null;
    }

    public boolean canAccept(MinoPosition minos) {
        for (Point p : minos.getMinos()) {
            if (!isEmptyAt(p.y, p.x)) {
                return false;
            }
        }
        return true;
    }

    public void put(MinoPosition position, PieceLabel k) {
        for (Point p : position.getMinos()) {
            cells[p.y][p.x] = k;
        }
        putHistory.add(k);
    }

    public void print() {
        for (PieceLabel[] row : cells) {
            StringBuilder sb = new StringBuilder();
            for (PieceLabel c : row) {
                sb.append(c == null ? "." : c.name()).append(" ");
            }
            System.out.println(colorString(sb.toString()));
        }
    }

    private static String colorString(String plainString) {
        final String RESET = "\033[0m";
        final Map<PieceLabel, String> colorMap = new EnumMap<>(PieceLabel.class);
        colorMap.put(PieceLabel.O, "\033[31m"); // RED
        colorMap.put(PieceLabel.P, "\033[32m"); // GREEN
        colorMap.put(PieceLabel.Q, "\033[33m"); // YELLOW
        colorMap.put(PieceLabel.R, "\033[34m"); // BLUE
        colorMap.put(PieceLabel.S, "\033[35m"); // MAGENTA
        colorMap.put(PieceLabel.T, "\033[36m"); // CYAN
        colorMap.put(PieceLabel.U, "\033[91m"); // LIGHT_RED
        colorMap.put(PieceLabel.V, "\033[92m"); // LIGHT_GREEN
        colorMap.put(PieceLabel.W, "\033[93m"); // LIGHT_YELLOW
        colorMap.put(PieceLabel.X, "\033[94m"); // LIGHT_BLUE
        colorMap.put(PieceLabel.Y, "\033[95m"); // LIGHT_MAGENTA
        colorMap.put(PieceLabel.Z, "\033[96m"); // LIGHT_CYAN

        StringBuilder coloredString = new StringBuilder();
        for (String s : plainString.split(" ")) {
            try {
                PieceLabel label = PieceLabel.valueOf(s);
                if (colorMap.containsKey(label)) {
                    coloredString.append(colorMap.get(label)).append(s).append(RESET).append(" ");
                } else {
                    coloredString.append(s).append(" ");
                }
            } catch (IllegalArgumentException e) {
                coloredString.append(s).append(" ");
            }
        }
        return coloredString.toString();
    }

    public Board deepCopy() {
        Board newBoard = new Board(getNy(), getNx());
        for (int i = 0; i < getNy(); i++) {
            for (int j = 0; j < getNx(); j++) {
                newBoard.cells[i][j] = this.cells[i][j];
            }
        }
        newBoard.putHistory.addAll(this.putHistory);
        return newBoard;
    }
}

class Piece {
    private static final int PENTO = 5;
    private final PieceLabel label;
    private final List<Point> minos;

    public Piece(PieceLabel label) {
        this.label = label;
        int[][] shape = PieceShape.shapes.get(label);
        this.minos = new ArrayList<>();
        for (int y = 0; y < shape.length; y++) {
            for (int x = 0; x < shape[y].length; x++) {
                if (shape[y][x] != 0) {
                    this.minos.add(new Point(x, y));
                }
            }
        }
    }

    public PieceLabel getLabel() {
        return label;
    }

    public List<MinoPosition> possiblePositions(int y, int x, Board board) {
        List<MinoPosition> positions = new ArrayList<>();
        List<Integer> turnFactors = new ArrayList<>();
        for (int i = 0; i < 8; i++) {
            turnFactors.add(i);
        }
        Collections.shuffle(turnFactors);

        List<Integer> cellIndexes = new ArrayList<>();
        for (int i = 0; i < PENTO; i++) {
            cellIndexes.add(i);
        }
        Collections.shuffle(cellIndexes);

        Set<List<Point>> done = new HashSet<>();
        for (int factor : turnFactors) {
            for (int centerCell : cellIndexes) {
                MinoPosition position = new MinoPosition(this.minos);
                position.turnFlip(factor);
                position.move(centerCell, y, x);

                List<Point> key = position.getMinos();
                if (done.contains(key)) {
                    continue;
                }
                done.add(key);

                if (board.canAccept(position)) {
                    positions.add(position);
                }
            }
        }
        return positions;
    }
}

class FitStatistics {
    private final Map<PieceLabel, Integer> trial = new EnumMap<>(PieceLabel.class);
    private final Map<PieceLabel, Integer> fit = new EnumMap<>(PieceLabel.class);

    public FitStatistics(List<Piece> pieces) {
        for (Piece p : pieces) {
            trial.put(p.getLabel(), 1);
            fit.put(p.getLabel(), 1);
        }
    }

    public void countupTrial(PieceLabel piece) {
        trial.put(piece, trial.get(piece) + 1);
    }

    public void countupFit(PieceLabel piece) {
        fit.put(piece, fit.get(piece) + 1);
    }

    public double fitRate(PieceLabel piece) {
        return (double) fit.get(piece) / trial.get(piece);
    }

    public List<Map.Entry<PieceLabel, Double>> listFitRate() {
        List<Map.Entry<PieceLabel, Double>> rateList = new ArrayList<>();
        for (PieceLabel p : trial.keySet()) {
            rateList.add(Map.entry(p, fitRate(p)));
        }
        rateList.sort(Map.Entry.comparingByValue());
        return rateList;
    }

    public double meanFitRate() {
        return (double) fit.values().stream().mapToInt(Integer::intValue).sum() / trial.values().stream().mapToInt(Integer::intValue).sum();
    }
}

class SearchStatistics {
    private final Map<Integer, Integer> reachedDepthCount = new HashMap<>();

    public void countup(int reachedDepth) {
        reachedDepthCount.put(reachedDepth, reachedDepthCount.getOrDefault(reachedDepth, 0) + 1);
    }

    public int totalTrial() {
        return reachedDepthCount.values().stream().mapToInt(Integer::intValue).sum();
    }

    public double meanReachedDepth() {
        return (double) reachedDepthCount.entrySet().stream().mapToLong(e -> (long) e.getKey() * e.getValue()).sum() / totalTrial();
    }

    public List<Map.Entry<Integer, Double>> reachDistribution() {
        List<Map.Entry<Integer, Double>> reachd = new ArrayList<>();
        long total = totalTrial();
        for (Map.Entry<Integer, Integer> entry : reachedDepthCount.entrySet()) {
            reachd.add(Map.entry(entry.getKey(), (double) entry.getValue() / total));
        }
        reachd.sort(Map.Entry.comparingByKey());
        return reachd;
    }
}

interface PieceChoiceStrategy {
    Piece pop(List<Piece> pieces, FitStatistics fitStatistics);
}

class FitDifficultPieceChoice implements PieceChoiceStrategy {
    @Override
    public Piece pop(List<Piece> pieces, FitStatistics fitStatistics) {
        pieces.sort(Comparator.comparingDouble(p -> fitStatistics.fitRate(p.getLabel())));
        return pieces.remove(pieces.size() - 1);
    }
}

class FitEasyPieceChoice implements PieceChoiceStrategy {
    @Override
    public Piece pop(List<Piece> pieces, FitStatistics fitStatistics) {
        pieces.sort(Comparator.comparingDouble((Piece p) -> fitStatistics.fitRate(p.getLabel())).reversed());
        return pieces.remove(pieces.size() - 1);
    }
}

class RandomPieceChoice implements PieceChoiceStrategy {
    @Override
    public Piece pop(List<Piece> pieces, FitStatistics fitStatistics) {
        Collections.shuffle(pieces);
        return pieces.remove(pieces.size() - 1);
    }
}
